Chapter10/pretty_print.erl

```diff
@@ -0,0 +1,194 @@
+		eof -> ok;
+		Line ->
+			processLine(Line,N),
+			processLines(IoDevice,N+1)
+	end.
+
+
+% -define(Punctuation,"(\\ |\\,|\\.|\\;|\\:|\\t|\\n|\\(|\\))+").
+-define(Punctuation, "[\\ |\\,|\\.|\\;|\\:|\\t|\\n|\\(|\\)]").
+
+processLine(Line,N) ->
+	Words = re:split(Line, ?Punctuation, [{return,list}]),
+	processWords(Words, N).
+
+
+processWords(Words,N) ->
+	case Words of
+		[] -> ok;
+		[Word|Rest] ->
+			if
+				length(Word) > 3 ->
+					Normalise = string:to_lower(Word),
+					ets:insert(indexTable, {{Normalise , N}} );
+				true -> ok
+			end,
+			processWords(Rest,N)
+	end.
+
+
+
+prettyIndex() ->
+	case ets:first(indexTable) of
+		'$end_of_table' -> ok;
+		First ->
+			case First of
+				{Word, N} ->
+					IndexEntry = {Word, [N]}
+			end,
+		prettyIndexNext(First, IndexEntry)
+	end.
+
+
+
+prettyIndexNext(Entry, {Word, Lines}=IndexEntry) ->
+	Next = ets:next(indexTable,Entry),
+	case Next of
+		'$end_of_table' -> prettyEntry(IndexEntry);
+		{NextWord, M} ->
+			if
+				NextWord == Word ->
+					prettyIndexNext(Next, {Word, [M|Lines]});
+				true ->
+					prettyEntry(IndexEntry),
+					prettyIndexNext(Next, {NextWord, [M]})
+			end
+	end.
+
+
+
+
+
+
+
+
+
+% Exercise 10-1: Pretty-Printing
+prettyEntry({Word, Lines}) when length(Lines) == 1 ->
+	Paded = pad(20, Word),
+	[Entry] = Lines,
+
+	Output = Paded ++ char_to_integer(Entry),
+
+	io:format("~p~n",[Output]);
+
+prettyEntry({Word, Lines}) ->
+	Acc = accumulate(Lines),
+	PList = prettyList(Acc),
+
+	Paded = pad(20, Word),
+
+	Output = Paded ++ PList,
+
+	io:format("~p~n",[Output]).
+
+
+
+
+
+% This function should take a list of line numbers,
+% in descending order, and produce a list containing
+% ranges as well as removing duplicates.
+accumulate(List) ->
+	Sorted = lists:sort(List),
+	% Compressed = compress(Sorted).
+	compress_link(Sorted).
+
+
+% This function will print the output of accumulate so
+% that on the list [{1},{3}, {5,7}] the output is 1,3,5-7
+prettyList(List) ->
+	compress_pretty(List).
+
+
+% This function, called with number N and string Word,
+% will return the string padded to the right with spaces
+% to give it length N (assuming Word is not longer than N).
+pad(N, Word) when length(Word) < N -> pad(N, Word ++ [$\s]);
+pad(_, Word) -> Word.
+
+
+
+
+%%Compresser (Removes duplicates) [1,1,1,2,2,5,7,9,9] -> [1,2,5,7,9]
+compress([]) -> [];
+compress(List) -> compress(List, []).
+
+compress([], Output) -> Output;
+compress(List, Output) when length(List) == 1 -> Output ++ List;	
+compress([Head|Tail], _) when (Tail - Head) =< 1 -> [];
+compress([Head|Tail], Output) when (Tail - Head) >= 1 -> Output ++ [Head];
+compress([Head,Next|Tail], Output) ->
+	if
+		(Next - Head) == 0 -> compress([Head|Tail], Output);
+		% (Next - Head) == 1 -> compress([Next|Tail], Output ++ [Head] ++ ["-"] );
+		(Next - Head) >= 1 -> compress([Next|Tail], Output ++ [Head])
+	end.
+
+
+%%Linker (Links adjacent numbers) [1,2,5,8,9,10] -> [{1,2},{5},{8,10}]
+compress_link([]) -> [];
```

Comment:
This implementation doesn't look readable and, to my mind.
Think how you can do it using recursion, break down each element you're processing into several cases, it makes implementation much easier to read (in my opinion).

```erlang
accumulate(List) ->
    accumulate(List, []).

accumulate([] = _Lines, Acc) ->
    Acc;
accumulate([Index | Tail], [] = _Acc) ->
    accumulate(Tail, [{Index}]);
accumulate([Index | Tail], [{Index} | _AccTail] = Acc) ->
    accumulate(Tail, Acc);
accumulate([Index | Tail], [{Index, _End} | _AccTail] = Acc) ->
    accumulate(Tail, Acc);
accumulate([Index | Tail], [{LineNum} | AccTail])
  when LineNum - 1 == Index ->
    accumulate(Tail, [{Index, LineNum} | AccTail]);
accumulate([Index | Tail], [{Start, End} | AccTail])
  when Start - 1 == Index ->
    accumulate(Tail, [{Index, End} | AccTail]);
accumulate([Index | Tail], Acc) ->
    accumulate(Tail, [{Index} | Acc]).
```

Reply:
Just updated the code to use pattern matching and guard tests instead of if statement, not sure about it being easier to read but it's probably a lot more efficient

Reply:
Sorry if I didn't explain it properly. My comment was not only about removing if statements and rewriting it using func guards, but rather about code complexity.
If you compare my implementation you can notice that there's only 1 arithmetic operation used (2 lines, but the same operation), which checks if element is adjoining or not.  In you code i see 2 different operation (_7_ lines), which makes code harder to read because you need to keep in mind all of those cases.
Plus you "split" function on processing elements, which there's something else left in the list and only when 2 last numbers left. Note that when you match on `[Head, Next | Tail]` it matches `[1, 2]` with `Tail` to be bind to empty list because [1,2] = [1, 2 | []]. So I'd say your implementation contains some extra lines, which are not really used or can be removed, which could reduce number of lines (and in this case would increase readability).